source/game/system/map/MapdataCheckPoint.cc

```diff
@@ -0,0 +1,258 @@
+}
+
+MapdataCheckPoint *MapdataCheckPoint::nextPoint(s32 i) const {
+    return m_nextPoints[i].checkpoint;
+}
+
+const LinkedCheckpoint &MapdataCheckPoint::nextLinked(s32 i) const {
+    return m_nextPoints[i];
+}
+
+/// @brief computes @ref m_prevKcpId for each checkpoint.
+/// @details @ref isPlayerFlagged indicates that a checkpoint has been visited, to prevent infinite
+/// recursion.
+/// https://decomp.me/scratch/MtF18
+/// @addr{0x80515A6C}
+void MapdataCheckPoint::linkPrevKcpIds(u8 prevKcpId) {
+    m_prevKcpId = isNormalCheckpoint() ? prevKcpId : this->type();
+
+    setPlayerFlags(0);
+    for (size_t i = 0; i < nextCount(); i++) {
+        MapdataCheckPoint *next = nextPoint(i);
+        if (next->isPlayerFlagged(0)) {
+            continue;
+        }
+        next->linkPrevKcpIds(m_prevKcpId);
+    }
+}
+
+/// @brief Returns true if player is between the two sides of the checkpoint quad, otherwise false
+/// @addr{Inlined in 0x80510C74}
+bool MapdataCheckPoint::checkSector(const LinkedCheckpoint &next, const EGG::Vector2f &p0,
+        const EGG::Vector2f &p1) const {
+    if (-(next.p0diff.y) * p0.x + next.p0diff.x * p0.y < 0.0f) {
+        return false;
+    }
+
+    if (next.p1diff.y * p1.x - next.p1diff.x * p1.y < 0.0f) {
+        return false;
+    }
+
+    return true;
+}
+
+/// @brief Updates @param checkpointCompletion\, which is the percentage of distance
+/// the player has traveled through the checkpoint quad
+/// @return True if 0 <= checkpointCompletion <= 1, meaning the player is between the current
+/// checkpoint line and the next; otherwise, false
+/// @addr{Inlined in 0x80510C74}
+bool MapdataCheckPoint::checkCheckpointCompletion(const LinkedCheckpoint &next,
+        const EGG::Vector2f &p0, const EGG::Vector2f &p1, float *checkpointCompletion) const {
+    f32 d1 = m_dir.dot(p1);
+    f32 d2 = -(next.checkpoint->m_dir.dot(p0));
+    // This is where the divide by zero thing happens @todo
+    f32 checkpointCompletion_ = d1 / (d1 + d2);
+    *checkpointCompletion = checkpointCompletion_;
+    return (checkpointCompletion_ >= 0.0f && checkpointCompletion_ <= 1.0f);
+}
+
+/// @brief Calls both @ref checkSector and @ref checkCheckpointCompletion; updates @param
+/// checkpointCompletion
+/// @addr{0x80510C74}
+MapdataCheckPoint::SectorOccupancy MapdataCheckPoint::checkSectorAndCheckpointCompletion_(
+        const LinkedCheckpoint &next, const EGG::Vector2f &p0, const EGG::Vector2f &p1,
+        float *checkpointCompletion) const {
+    if (!checkSector(next, p0, p1)) {
+        return OutsideSector;
+    }
+
+    return checkCheckpointCompletion(next, p0, p1, checkpointCompletion) ?
+            InsideSector :
+            OutsideSector_BetweenSides;
+}
+
+/// @addr{0x80512064}
+f32 MapdataCheckPointAccessor::calculateMeanTotalDistanceRecursive(u16 ckptId) {
+    f32 sumDist = 0.0f;
+    MapdataCheckPoint *ckpt = get(ckptId);
+    u16 n = ckpt->nextCount();
+    for (size_t i = 0; i < n; i++) {
```

Comment:
Stylistic choice but can you add newlines before/after for loops and if-statements? It helps break up the flow of your code